File: fabric-samples/chaincode/fly/go/flynet.go
package main

import (
	"encoding/json"
	"strconv"
	"errors"
	"fmt"

	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// FlyNetChaincode implements the chaincode interface for FlyNet network
type FlyNetChaincode struct {
	contractapi.Contract
}

// FlyNetChaincodeResult structure used for handling result of query
type FlyNetChaincodeResult struct {
	Key		string `json:"Key"`
	Record *FlyNetChaincode
}

// Flight represents a flight details of what makes up a flight
type Flight struct {
	FlyFrom				string `json:"flyFrom"`
	FlyTo				string `json:"flyTo"`
	dateTime			string `json:"dateTime"`
	availablePlaces		int    `json:"availablePlaces"`
	BookingStatuses		map[string]string `json:"bookingStatus"`
}

// Reservation represents a reservation of a flight
type Reservation struct {
	ReservationNr		string `json:"reservationNr"`
	CustomerNames		[]string `json:"customerNames"`
	customerEmail		string `json:"customerEmail"`
	FlightNr			string `json:"flightNr"`
	NrOfSeats			int    `json:"nrOfSeats"`
	Status 				string `json:"status"`
}

// Generates flightNr
func generateFlightNr(Org number) string {
	if Org == "Org1MSP" {
		return "EC" + strconv.Itoa(rand.Intn(1000))
	}
	if Org == "Org3MSP" {
		return "BS" + strconv.Itoa(rand.Intn(1000))
	}
	return nil
}

// CreateFlight creates a new flight on the ledger
func (t *FlyNetChaincode) CreateFlight(ctx contractapi.TransactionContextInterface, flyFrom string, flyTo string, dateTime string, seats int) (string, error) {
	callerOrg, err := ctx.GetClientIdentity().GetMSPID()
	
	if err != nil {
		return err, fmt.Errorf("Failed getting caller's org: %v", err)
	}

	if callerOrg != "Org1MSP" || callerOrg != "Org3MSP" {
		return err, fmt.Errorf("The caller is not authorized to invoke the CreateFlight function")
	}

	flightNr, err := generateFlightNr(callerOrg)

	if err == nil {
		return err, fmt.Errorf("Failed generating flightNr: %v", err)
	}
	
	flight := Flight{
		FlyFrom:			flyFrom,
		FlyTo:				flyTo,
		dateTime:			dateTime,
		availablePlaces:	seats,
	}

	flightAsBytes, err := json.Marshal(flight)
	if err != nil {
		return err, fmt.Errorf("Failed converting flight to bytes: %v", err)
	}

	err = ctx.GetStub().PutState(flightNr, flightAsBytes)
	if err != nil {
		return err, fmt.Errorf("Failed to put flight on ledger: %v", err)
	}
	return "Congrats, flight was created successfully", nil
}

File: fabric-samples/chaincode/fly/js/flynet.js
'use strict';

const { Contract } = require('fabric-contract-api');

let lastFlightNr = 0;
let lastReservationNr = 0;

class FlyNetChaincode extends Contract {
    async createFlight(ctx, flyFrom, flyTo, dateTime, seats) {
        const callerOrg = ctx.clientIdentity.getMSPID();

        if (callerOrg !== 'Org1MSP' && callerOrg !== 'Org3MSP') {
            throw new Error('The caller is not authorized to invoke the CreateFlight function');
        }

        const flightNr = this.generateFlightNr(callerOrg);
        lastFlightNr += 1;

        const flight = {
            ID: flightNr,
            FlyFrom: flyFrom,
            FlyTo: flyTo,
            dateTime: dateTime,
            availablePlaces: seats,
            BookingStatuses: {}
        };

        await ctx.stub.putState(flightNr, Buffer.from(JSON.stringify(flight)));

        return 'Congrats, flight was created successfully';
    }

    async getAllFlights(ctx) {
        const startKey = '';
        const endKey = ''; // the maximum unicode value

        const iterator = await ctx.stub.getStateByRange(startKey, endKey);
        const results = [];

        let res = await iterator.next();
        while (!res.done) {
            const item = {
                key: res.value.key,
                value: JSON.parse(res.value.value.toString('utf8'))
            };
            try {
                const is = item.value.flightNr;
                results.push(item);
            } catch (err) {
            }

            res = await iterator.next();
        }

        await iterator.close();
        return results;

    }
    async getFlight(ctx, flightNr) {
        const flightAsBytes = await ctx.stub.getState(flightNr);

        if (!flightAsBytes || flightAsBytes.length === 0) {
            throw new Error(`Flight ${flightNr} does not exist`);
        }

        const flight = JSON.parse(flightAsBytes.toString('utf8'));
        return flight;
    }

    async reserveSeats(ctx, flightNr, number, customerNames, customerEmail) {
        const callerOrg = ctx.clientIdentity.getMSPID();

        if (callerOrg !== 'Org2MSP') {
            throw new Error('The caller is not authorized to invoke the ReserveSeats function');
        }

        const flight = await this.getFlight(ctx, flightNr);

        if (Number(flight.availablePlaces) < Number(number)) {
            throw new Error('Not enough available places');
        }

        flight.availablePlaces -= number;

        await ctx.stub.putState(flightNr, Buffer.from(JSON.stringify(flight)));

        const reservationNr = 'R' + lastReservationNr;
        lastReservationNr += 1;

        const reservation = {
            ID: reservationNr,
            CustomerNames: customerNames,
            CustomerEmail: customerEmail,
            FlightNr: flightNr,
            NrOfSeats: number,
            Status: 'Pending'
        };

        await ctx.stub.putState(reservationNr, Buffer.from(JSON.stringify(reservation)));

        return 'Reservation created' + reservationNr;
    }

    async getReservation(ctx, reservationNr) {
        const reservationAsBytes = await ctx.stub.getState(reservationNr);

        if (!reservationAsBytes || reservationAsBytes.length === 0) {
            throw new Error(`Reservation ${reservationNr} does not exist`);
        }

        const reservation = JSON.parse(reservationAsBytes.toString('utf8'));
        return reservation;
    }

    async bookSeats(ctx, reservationNr) {
        const callerOrg = ctx.clientIdentity.getMSPID();

        if (callerOrg !== 'Org1MSP' && callerOrg !== 'Org3MSP') {
            throw new Error('The caller is not authorized to invoke the BookSeats function');
        }

        const reservation = await this.getReservation(ctx, reservationNr);
        const flight = await this.getFlight(ctx, reservation.FlightNr);

        const airlinePrefix = flight.ID.substring(0, 2);
        if ((airlinePrefix === 'EC' && callerOrg !== 'Org1MSP') || (airlinePrefix === 'BS' && callerOrg !== 'Org3MSP')) {
            throw new Error('The caller is not authorized to book seats for this flight');
        }
        reservation.Status = 'Completed';

        await ctx.stub.putState(reservationNr, Buffer.from(JSON.stringify(reservation)));

        return 'Reservation confirmed';
    }

    async checkIn(ctx, reservationNr, passportIDs) {
        const callerOrg = ctx.clientIdentity.getMSPID();

        if (callerOrg !== 'Org2MSP') {
            throw new Error('The caller is not authorized to invoke the CheckIn function');
        }

        const reservation = await this.getReservation(ctx, reservationNr);
        const flight = await this.getFlight(ctx, reservation.FlightNr);

        // Check if passport IDs match the customerNames in the reservation
        if (passportIDs.length !== reservation.CustomerNames.length) {
            throw new Error('The number of passport IDs does not match the number of customers');
        }

        reservation.CustomerNames.forEach((customerName, index) => {
            if (customerName !== passportIDs[index]) {
                throw new Error('The passport IDs do not match the customer names');
            }
        });

        //Chage status to checked-in
        reservation.Status = 'CheckedIn';

        await ctx.stub.putState(reservationNr, Buffer.from(JSON.stringify(reservation)));

        // Json object of tickets
        const tickets = {
            ReservationNr: reservationNr,
            CustomerNames: reservation.CustomerNames,
            CustomerEmail: reservation.CustomerEmail,
            FlightNr: reservation.FlightNr,
            FlyFrom: flight.FlyFrom,
            FlyTo: flight.FlyTo,
            dateTime: flight.dateTime,
            NrOfSeats: reservation.NrOfSeats
        };

        return tickets;
    }

    generateFlightNr(org) {
        if (org === 'Org1MSP') {
            return 'EC' + lastFlightNr;
        } else {
            return 'BS' + lastFlightNr;
        }
    }
}

module.exports = FlyNetChaincode;